Simplified/NYPLUserAccount.swift

```diff
@@ -10,21 +10,141 @@ extension Notification.Name {
-  private var deviceIDKey = "NYPLAccountDeviceIDKey"
+  private lazy var keychainTransaction = KeychainVariableTransaction(accountInfoLock: accountInfoLock)
+    
+  private var libraryUUID: String? {
+    didSet {
+      guard libraryUUID != oldValue else { return }
+      let variables: [StorageKey: Keyable] = [
+        StorageKey.authorizationIdentifier: _authorizationIdentifier,
+        StorageKey.adobeToken: _adobeToken,
+        StorageKey.licensor: _licensor,
+        StorageKey.patron: _patron,
+        StorageKey.adobeVendor: _adobeVendor,
+        StorageKey.provider: _provider,
+        StorageKey.userID: _userID,
+        StorageKey.deviceID: _deviceID,
+        StorageKey.credentials: _credentials,
+        StorageKey.authDefinition: _authDefinition,
+        StorageKey.cookies: _cookies,
+
+        // legacy
+        StorageKey.barcode: _barcode,
+        StorageKey.PIN: _pin,
+        StorageKey.authToken: _authToken,
+      ]
+
+      for (key, var value) in variables {
+        value.key = key.keyForLibrary(uuid: libraryUUID)
+      }
+//      removeAll() // to clean keychain data for a library <- REMOVE ME
+    }
+  }
+
+  var authDefinition: AccountDetails.Authentication? {
+    get {
+      let legacyDefinition: AccountDetails.Authentication?
+      if let libraryUUID = self.libraryUUID {
+        legacyDefinition = AccountsManager.shared.account(libraryUUID)?.details?.auths.first
+      } else {
+        legacyDefinition = AccountsManager.shared.currentAccount?.details?.auths.first
+      }
+      return _authDefinition.read() ?? legacyDefinition
+    }
+    set {
+      guard let newValue = newValue else { return }
+      _authDefinition.write(newValue)
+
+      DispatchQueue.main.async {
+        var mainFeed = URL(string: AccountsManager.shared.currentAccount?.catalogUrl ?? "")
+        let resolveFn = {
+          NYPLSettings.shared.accountMainFeedURL = mainFeed
+          UIApplication.shared.delegate?.window??.tintColor = NYPLConfiguration.mainColor()
+          NotificationCenter.default.post(name: NSNotification.Name.NYPLCurrentAccountDidChange, object: nil)
+        }
+
+        if self.needsAgeCheck {
+          AgeCheck.shared().verifyCurrentAccountAgeRequirement { [weak self] meetsAgeRequirement in
+            DispatchQueue.main.async {
+              mainFeed = meetsAgeRequirement ? self?.authDefinition?.coppaOverUrl : self?.authDefinition?.coppaUnderUrl
+              resolveFn()
+            }
+          }
+        } else {
+          resolveFn()
+        }
+      }
+
+      notifyAccountDidChange()
+    }
+  }
+
+  public private(set) var credentials: Credentials? {
+    get {
+      var credentials = _credentials.read()
+
+      if credentials == nil {
+        // try to load legacy values
+        if let barcode = legacyBarcode, let pin = legacyPin {
+          credentials = .barcodeAndPin(barcode: barcode, pin: pin)
+
+          keychainTransaction.perform {
+            _credentials.write(credentials)
```

Comment:
I added a description why is there a write operation, about the userID, frankly speaking I don't know. It's related to Adept which I couldn't use and test

Reply:
the setUserID is not related to Adept: it's setting the (hashed) user id for Crashlytics reporting, so that we can easily track errors for a particular user. I think it's something we should preserve, or are you saying you don't have a barcode in this case?

Reply:
I don't think there is any way of tracking a user in SAML. There's no barcode obviously and I don't recall any identifier that would be suitable for this matter

Reply:
sounds good, thanks for clarifying.